pkg/provider/helm/release.go

```diff
@@ -0,0 +1,331 @@
+package helm
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/mcuadros/go-defaults"
+	"helm.sh/helm/v3/pkg/action"
+	"helm.sh/helm/v3/pkg/chart"
+	"helm.sh/helm/v3/pkg/chart/loader"
+	"helm.sh/helm/v3/pkg/release"
+)
+
+var ErrReleaseNotFound = errors.New("release not found")
+var ErrChartNotApplication = errors.New("helm chart is not an application chart")
+
+type releaseConfig struct {
+	// Name - Release Name
+	Name string `valid:"required"`
+	// Repository - Repository where to locate the requested chart. If is a URL the chart is installed without installing the repository.
+	Repository string `valid:"required"`
+	// Chart - Chart name to be installed. A path may be used.
+	Chart string `valid:"required"`
+	// Version - Specify the exact chart version to install. If this is not specified, the latest version is installed.
+	Version string
+	// Values - Map of values in to pass to helm.
+	Values map[string]interface{}
+	// Namespace - Namespace to install the release into.
+	Namespace string `default:"default"`
+	// Timeout - Time in seconds to wait for any individual kubernetes operation.
+	Timeout int `default:"300"`
+	// ForceUpdate - Force resource update through delete/recreate if needed.
+	ForceUpdate bool `default:"false"`
+	// RecreatePods - Perform pods restart during upgrade/rollback
+	RecreatePods bool `default:"false"`
+	// Wait - Will wait until all resources are in a ready state before marking the release as successful.
+	Wait bool `default:"true"`
+	// WaitForJobs - If wait is enabled, will wait until all Jobs have been completed before marking the release as successful.
+	WaitForJobs bool `default:"false"`
+	// Replace - Re-use the given name, even if that name is already used. This is unsafe in production
+	Replace bool `default:"false"`
+	// Description - Add a custom description
+	Description string
+	// CreateNamespace - Create the namespace if it does not exist
+	CreateNamespace bool `default:"false"`
+}
+
+func DefaultReleaseConfig() *releaseConfig {
+	defaultReleaseConfig := new(releaseConfig)
+	defaults.SetDefaults(defaultReleaseConfig)
+	return defaultReleaseConfig
+}
+
+type Release struct {
+	Config *releaseConfig
+	Output ReleaseOutput
+}
+
+type ReleaseOutput struct {
+	// Status - Status of the release.
+	Status Status
+	// Revision - Revision of the release.
+	Release string
+}
+
+// Release - creates or updates a helm release with its configs
+func (p *Provider) Release(config *releaseConfig) (*Release, error) {
+	releaseExists, _ := p.resourceReleaseExists(config.Name, config.Namespace)
+	if releaseExists {
+		return p.update(config)
+	}
+	return p.create(config)
+}
+
+func (p *Provider) create(config *releaseConfig) (*Release, error) {
+	actionConfig, err := p.getActionConfiguration(config.Namespace)
+	if err != nil {
+		return nil, err
```

Comment:
Let's wrap these errors into our own error. It will make it easier to handle errors where we use this methods.